guide: ML Pipelines (1): Defining Pipelines & Stages (#3414)
iterative/dvc.org: content/docs/user-guide/pipelines/index.md

---
title: Pipelines Management in DVC
---

> # What is a (DVC) pipeline?

Reviewer: Also we need this doc in sidebar.json 😬

> # What is a (DVC) pipeline?

Author: Updated the title. Added to the sidebar, just below experiment management.

> # What is a (DVC) pipeline?

Reviewer: Still seeing a mismatch: It's just **Pipelines** in nav, **DVC Pipelines** in the frontmatter, and **What is a (DVC) pipeline?** in the HTML.

> A machine learning pipeline is like an assembly line in a factory. As raw

Reviewer: I like the analogy but it's a bit general i.e. any pipeline is like an assembly line. What makes an ML pipeline special?

Author: "Pipeline" in software is usually shell pipes, like `ls -t | head -n 3`. "Assembly line" analogy is to counter that actually. I think if we start from the general pipelines, we may need to touch the shell pipes to avoid confusion. This was just to say "pipelines we talk in ML are not pipelines in shell."

Reviewer: If we call the document **ML Pipelines** then we don't have to disambiguate here 🙂

materials input to a factory and processed at each step to achieve a final
product, a machine learning pipeline gets the data as an input, processes it in
every stage and outputs a model.

If you are repeating a set of commands to get to the final artifacts (models,
results), then you already have a pipeline, albeit in a manual fashion. DVC

> allows to automate your workflow, be it comprised of a single command or many

Reviewer: I think it's not so much about manual vs automatic. You can automate any set of commands with a bash script.

DVC codifies it into a readable, std. format + provides commands to run it (which are manual too).

Reviewer: Feel free to double check with the DVC team cc @dberenbaum

Author: Reproduction can't work without codification first. We don't have functionality to find the relationships between stages automatically.

commands with complex relationships.

## Stages

> A pipeline is a collection of stages. At each stage, we define a (shell) command

Reviewer: ```suggestion
A DVC pipeline is a collection of stages. At each stage, we define a (shell) command
```

> A pipeline is a collection of stages. At each stage, we define a (shell) command
> to run and specify the inputs and outputs for this stage. By defining stage
> input and outputs, we can find the order to run the stages. If an output of
> stage `A` is fed as an input to stage `B`, then DVC infers to run `A` before
> `B`.

Reviewer: Proposed clarifications
```suggestion
A DVC pipeline is a collection of connected <abbr>stages</abbr>. At each stage, we define a shell command
to run, and specify its inputs and outputs (if any). By defining
outputs that feed into future inputs, we can determine the stages' execution order, e.g. if an output of
stage X is fed as an input to stage Y, then DVC infers to run X before
Y.
```

`dvc stage` set of commands are used to create a pipeline by defining its
stages. Each stage requires a `name`, and a `command`. Additionally it describes
a set of dependencies and outputs. When these outputs are missing or the

> dependencies are newer than the outputs, the command is run.

Reviewer: This paragraph seems mostly redundant, esp. the part about defining inputs and outputs. `dvc stage` "set" is just 2 commands so probably no need to mention it (go straight to `stage add`, already in the next p). The last sentence "When these outputs are missing " could be included in the previous paragraph that describes stages, or even better later to the text about execution (after dvc.yaml sample).

There are two equally valid ways to describe stages. The first is using

> `dvc stage add` command, and the other is by editing `dvc.yaml`.

Reviewer: I think we want to mention dvc.yam first -- encourage users to write them manually.

Reviewer: ☝🏼 implies show the dvc.yaml sample first and then mention how it could be created with the `stage add` HELPER.

For example, to add a stage named `preprocess` depending on `preprocess.py`

```dvc
$ dvc stage add --name preprocess \
                --deps src/preprocess.py \
                --deps data/raw \
                --outs data/preprocessed \
                python src/preprocess.py
```

The other way to define a stage is through editing `dvc.yaml`. You could create
the file with the following content to create the above stage:

```yaml
stages:
  preprocess:
    cmd: python src/preprocess.py
    deps:
      - data/raw
      - src/preprocess.py
    outs:
      - data/preprocessed
```

Essentially the command creates the stage by writing to `dvc.yaml`file. If you
are creating the pipeline for the first time, adding the stages by
`dvc stage add` may be easier. If you're editing the pipeline though, working

> with the YAML file might be easier.

Reviewer: I'd summarize/incorporate into the paragraph before dvc.yaml sample

<admon type="tip">

> Some advanced configuration for the stages may only be available through
> editing `dvc.yaml`.
> </admon>
> 
> <admon type="warning">
> DVC works with YAML 1.2 and any version before this may have some minor quirks
> that cause subtle bugs.

Reviewer: Blank lines needed for proper rendering I think. Let me commit this so we get a Gatsby Cloud review app...

```suggestion

Some advanced configuration for the stages may only be available through
editing `dvc.yaml`.

</admon>

<admon type="warning">

DVC works with YAML 1.2 and any version before this may have some minor quirks
that cause subtle bugs.

```

dependencies are newer than the outputs, the command is run.

There are two equally valid ways to describe stages. The first is using
`dvc stage add` command, and the other is by editing `dvc.yaml`.

For example, to add a stage named `preprocess` depending on `preprocess.py`

```dvc
$ dvc stage add --name preprocess \
                --deps src/preprocess.py \
                --deps data/raw \
                --outs data/preprocessed \
                python src/preprocess.py
```

The other way to define a stage is through editing `dvc.yaml`. You could create
the file with the following content to create the above stage:

```yaml
stages:
  preprocess:
    cmd: python src/preprocess.py
    deps:
      - data/raw
      - src/preprocess.py
    outs:
      - data/preprocessed
```

Essentially the command creates the stage by writing to `dvc.yaml`file. If you
are creating the pipeline for the first time, adding the stages by
`dvc stage add` may be easier. If you're editing the pipeline though, working
with the YAML file might be easier.

<admon type="tip">

Some advanced configuration for the stages may only be available through
editing `dvc.yaml`.

</admon>

> <admon type="warning">

Reviewer: Also, the type is just `warn`, I think.

```suggestion
<admon type="warn">
```

DVC works with YAML 1.2 and any version before this may have some minor quirks
that cause subtle bugs.

> </admon>

Reviewer: May not need an admonition for this, just mention "YAML 1.2" in the paragraph that introduces dvc.yaml ?

## Directed-Acyclic Graph (DAG)

By adding stages to the pipeline, we define a [graph] where the nodes are stages

> and the edges are dependency relationships. The final topology of the graph

Reviewer: BTW we should probably also bring some text from https://dvc.org/doc/command-reference/dag into here and remove most of the explanations from that cmd ref. Rel. (as mentioned in #2883)

Reviewer: p.s. on this: we currently say that "pipelines are series of stages" in a few places (including the command ref which content will now be merged into here). Probably OK in the Get Started but the guide should be more nuanced: a pipeline can have serial or parallel shapes; it can be any DAG.

should be a [Directed Acyclic Graph]. We know why it's directed, stage A depends
on stage B is different from stage B depends on stage A. But why it should be
_acyclic_?

The pipeline graph shouldn't contain any cycles in the form
`A depends on B depends on C depends on A`. Otherwise, invalidating one of the
stages causes pipeline to run indefinitely. If we invalidate `B` in this
example, it will cause `A` to be run, and as `C` depends on `A`, it will be
invalidated and run too. Finally, as `B` depends on `C`, it will need be
invalidated again. It never ends, it's an infinite loop.

Because of this, DVC checks the stage dependencies for an existence of a cycle
in the graph, and ensures that the graph is a [DAG].

If you want to visualize the graph yourself, you can use `dvc dag` command. It
outputs a text representation of the stages by default.

```dvc
$ dvc dag
```

TODO: output of `dvc dag example`

It can also output the graph in DOT format, that you can supply to `dot` to get
an image.

```dvc
$ dvc dag --dot | dot -Tpng pipeline.png
```

TODO: output of `dvc dag --dot`

ASK: Tell `dvc dag --output` here or skip it?

> <admon type="warning">
> DVC assumes the command you specify in a stage changes the outputs you specify.
> Otherwise, the stage may never be validated and the pipeline will always run.
> </admon>

Reviewer: Apparently this renders fine but I remember there's occasions where no new line causes issues so let's just always use them for consistency 🙂 

```suggestion
<admon type="warning">

DVC assumes the command you specify in a stage changes the outputs you specify.
Otherwise, the stage may never be validated and the pipeline will always run.

</admon>
```

Reviewer: Minor: a few`<admon>` tags still lack spacing ☝🏼

## Dependencies

A stage has three different types of ingredients, command, dependencies, and

> outputs. Commands are a required part of `dvc stage add`, and the other two are

Reviewer: Not sure about the structure of the doc. Stages is a section, then DAG, and then stage dependencies?

This should be part of the Stages section or have a different focus prob.

optional. Although optional, most of the features that characterize pipelines
are brought forward by dependency definitions.

DVC has more than one type of dependency: The basic one, that we call only
`dependency` is either a file or a directory. A stage that depends on a file or
directory is invalidated when this file or directory _contents_ changes.

> <admon type="note">
> DVC doesn't only check the timestamp of files, it actually calculates the hash
> of their contents to invalidate the dependent stages. This is one of the
> distinctive features over other build tools like  `make`. 

Reviewer: Same
```suggestion
<admon type="note">

DVC doesn't only check the timestamp of files, it actually calculates the hash
of their contents to invalidate the dependent stages. This is one of the
distinctive features over other build tools like  `make`. 

```

---
title: DVC Pipelines
---

# What is a (DVC) pipeline?

A machine learning pipeline is like an assembly line in a factory. As raw
materials input to a factory and processed at each step to achieve a final
product, a machine learning pipeline gets the data as an input, processes it in
every stage and outputs a model.

If you are using a sequence of commands to get the final artifacts (models,
results), then you already have a pipeline, albeit manually. DVC allows to
automate your workflow, be it comprised of a single command or many commands
with complex relationships.

> ## Stages

Reviewer: Please link from https://dvc.org/doc/command-reference/stage to this section.

Reviewer: Also from https://dvc.org/doc/user-guide/glossary#stage

A DVC pipeline is a collection of connected <abbr>stages</abbr>. At each stage,
we define a shell command to run, and specify its inputs and outputs (if any).
By defining outputs that feed into future inputs, we can determine the stages'
execution order, e.g. if an output of stage X is used as an input in stage Y,

> then DVC infers to run X before Y.

Reviewer: Mention/link to DAG concept

  changes in these files but doesn't make any assumptions regarding their use.

- **Metrics**: DVC pipelines has special support for metrics files. These files
  contain model comparison metrics that are output after training, validation or
  testing phases of model production. DVC allows to compare produced metrics
  with one another using `dvc metrics diff` and presents the results as a table
  with `dvc metrics show`.

- **Plots**: A more specialized kind of metrics files can be used to create plot
  graphs. In contrast to metrics files, plots are used to present progression of
  more than one training step, like epochs in deep learning projects. DVC
  supports to supply plots files to Vega-Lite and present them on browser with
  `dvc plots show`.

- **Image Plots**: Another kind of output is plots produced by the commands.
  These are standard image files that can be shown on browsers.

### Plots Outputs

## Running Pipelines

- Why?
- How?

## Multiple Pipelines

## Experiments with Pipelines

DVC 2.0 introduced experiment management with `dvc exp` set of commands. We have
a dedicated part in the user's guide for [experiment management]. Here, we

> briefly touch the relationship between experiments and pipelines.

Reviewer: What's this section for?
